src/StmtToViz.cpp

```diff
@@ -0,0 +1,2892 @@
+        set_compute_costs(op, 0, {op->value.get()});
+        set_data_costs(op, 0, {op->value.get()});
+    }
+
+    void visit(const Atomic *op) override {
+        IRVisitor::visit(op);
+        set_compute_costs(op, 0, {op->body.get()}, {});
+        set_data_costs(op, 0, {op->body.get()}, {});
+    }
+};
+
+/******************* GetAssemblyInfo *******************/
+// Used to map some Halide IR nodes to line-numbers in the
+// assembly file containing the corresponding generated code.
+class AssemblyInfo : public IRVisitor {
+public:
+    AssemblyInfo()
+        : loop_id(0), prodcons_id(0) {
+    }
+
+    void generate(const std::string &code, const Module &m) {
+        // Traverse the module to populate the list of
+        // nodes we need to map and generate their assembly
+        // markers (comments that appear in the assembly code
+        // associating the code with this node)
+        for (const auto &fn : m.functions()) {
+            fn.body.accept(this);
+        }
+
+        // Find markers in asm code
+        std::istringstream asm_stream(code);
+        std::string line;
+        int lno = 1;
+        while (getline(asm_stream, line)) {
+            // Try all markers
+            std::vector<uint64_t> matched_nodes;
+            for (auto const &[node, regex] : markers) {
+                if (std::regex_search(line, regex)) {
+                    // Save line number
+                    lnos[node] = lno;
+                    // Save this node's id
+                    matched_nodes.push_back(node);
+                }
+            }
+            // We map to the first match, stop
+            // checking matched nodes
+            for (auto const &node : matched_nodes) {
+                markers.erase(node);
+            }
+
+            lno++;
+        }
+    }
+
+    int get_asm_lno(uint64_t node_id) {
+        if (lnos.count(node_id)) {
+            return lnos[node_id];
+        }
+        return -1;
+    }
+
+private:
+    // Generate asm markers for Halide loops
+    int loop_id;
+    int gen_loop_id() {
+        return ++loop_id;
+    }
+
+    std::regex gen_loop_asm_marker(int id, const std::string &loop_var) {
+        std::regex dollar("\\$");
+        std::string marker = "%\"" + std::to_string(id) + "_for " + loop_var;
+        marker = std::regex_replace(marker, dollar, "\\$");
+        return std::regex(marker);
+    }
+
+    // Generate asm markers for Halide producer/consumer ndoes
+    int prodcons_id;
+    int gen_prodcons_id() {
+        return prodcons_id;
+    }
+
+    std::regex gen_prodcons_asm_marker(int id, const std::string &var, bool is_producer) {
+        std::regex dollar("\\$");
+        std::string marker = "%\"" + std::to_string(id) + (is_producer ? "_produce " : "_consume ") + var;
+        marker = std::regex_replace(marker, dollar, "\\$");
+        return std::regex(marker);
+    }
+
+    // Mapping of IR nodes to their asm markers
+    std::map<uint64_t, std::regex> markers;
```

Comment:
Why is this a uint64_t?

Reply:
I used the pointer values as IDs to connect JavaScript with HTML elements

Reply:
So they show up as integers in the html source? That would make the generated html non-deterministic, which may interfere with downstream consumers of these files (e.g. if they ever get checked into source the diffs when they update will be pointlessly large). I'd feel better if there were a mapping from const IRNode * to unique ids instead.

Reply:
Yes they do. And fair point, this would be really bad for any sort of regression testing. I'll assign unique IDs based on traversal order instead.